File: lib/tools/result-checker.ts
import { adminSupabase } from '@/lib/supabase/admin'
import { getMatchResult } from '@/lib/tools/football-api'
import { updatePerformance } from '@/lib/tools/memory'
import { sendMessage, formatResultAnnouncement } from '@/lib/tools/telegram'
import type { Tier } from '@/lib/types'

interface PendingPick {
  id: string
  session_id: string
  home_team: string
  away_team: string
  competition: string
  match_datetime: string
  bet_type: string
  odds: number
  result: null
}

interface SessionWithPicks {
  id: string
  tier: Tier
  date: string
  combined_odds: number | null
  telegram_msg_id: string | null
  picks: Array<{ result: 'win' | 'loss' | 'void' | null; was_rejected: boolean }>
}

function evaluateResult(
  betType: string,
  homeGoals: number,
  awayGoals: number
): 'win' | 'loss' | 'void' {
  const total = homeGoals + awayGoals
  const bt = betType.toLowerCase()

  if (bt.includes('moins de 1.5') || bt.includes('under 1.5')) return total < 1.5 ? 'win' : 'loss'
  if (bt.includes('plus de 1.5') || bt.includes('over 1.5')) return total > 1.5 ? 'win' : 'loss'
  if (bt.includes('moins de 2.5') || bt.includes('under 2.5')) return total < 2.5 ? 'win' : 'loss'
  if (bt.includes('plus de 2.5') || bt.includes('over 2.5')) return total > 2.5 ? 'win' : 'loss'
  if (bt.includes('moins de 3.5') || bt.includes('under 3.5')) return total < 3.5 ? 'win' : 'loss'
  if (bt.includes('plus de 3.5') || bt.includes('over 3.5')) return total > 3.5 ? 'win' : 'loss'

  if (bt.includes('btts oui') || bt.includes('les deux équipes marquent'))
    return homeGoals > 0 && awayGoals > 0 ? 'win' : 'loss'
  if (bt.includes('btts non'))
    return homeGoals === 0 || awayGoals === 0 ? 'win' : 'loss'

  if (bt.includes('victoire') && bt.includes('domicile')) return homeGoals > awayGoals ? 'win' : 'loss'
  if (bt.includes('victoire') && bt.includes('extérieur')) return awayGoals > homeGoals ? 'win' : 'loss'
  if (bt.includes('nul') || bt.includes('draw')) return homeGoals === awayGoals ? 'win' : 'loss'

  // Victoire [nom équipe] — domicile si nommée, sinon void
  if (bt.startsWith('victoire ')) {
    const teamName = bt.replace('victoire ', '').trim()
    if (teamName && homeGoals > awayGoals) return 'win'
    if (teamName && awayGoals > homeGoals) return 'loss'
  }

  return 'void'
}

export async function checkPendingResults(): Promise<{ checked: number; updated: number }> {
  const now = new Date()
  const cutoff = new Date(now.getTime() - 2 * 60 * 60 * 1000).toISOString()

  const { data: pending } = await adminSupabase
    .from('picks')
    .select('id, session_id, home_team, away_team, competition, match_datetime, bet_type, odds, result')
    .is('result', null)
    .eq('was_rejected', false)
    .lt('match_datetime', cutoff)
    .limit(50)

  if (!pending?.length) return { checked: 0, updated: 0 }

  let updated = 0

  for (const pick of pending as PendingPick[]) {
    const matchDate = pick.match_datetime.split('T')[0]
    const score = await getMatchResult(pick.home_team, pick.away_team, matchDate)
    if (!score) continue

    const result = evaluateResult(pick.bet_type, score.home, score.away)

    await adminSupabase
      .from('picks')
      .update({ result, result_checked_at: new Date().toISOString() })
      .eq('id', pick.id)

    // Utilise la compétition directement depuis le pick (plus besoin de requête supplémentaire)
    const competition = pick.competition || 'Unknown'
    await updatePerformance(pick.bet_type, competition, result, pick.odds)
    updated++
  }

  return { checked: pending.length, updated }
}

/**
 * Poste le résultat d'un combiné sur Telegram une fois que TOUS ses picks
 * sont résolus (win/loss/void) — un seul pick perdant fait perdre tout le
 * combiné, comme un vrai pari combiné. Chaque session n'est annoncée
 * qu'une fois (result_posted_at). Appelé après checkPendingResults() dans
 * le même cron, pour que les picks soient déjà à jour.
 */
export async function announceSessionResults(): Promise<{ checked: number; announced: number }> {
  const { data: sessions } = await adminSupabase
    .from('pronostic_sessions')
    .select('id, tier, date, combined_odds, telegram_msg_id, picks(result, was_rejected)')
    .eq('status', 'published')
    .is('result_posted_at', null)
    .limit(20)

  if (!sessions?.length) return { checked: 0, announced: 0 }

  let announced = 0

  for (const session of sessions as unknown as SessionWithPicks[]) {
    const picks = session.picks.filter(p => !p.was_rejected)
    if (!picks.length) continue

    const allResolved = picks.every(p => p.result !== null)
    if (!allResolved) continue

    const hasLoss = picks.some(p => p.result === 'loss')
    const allVoid = picks.every(p => p.result === 'void')
    const comboResult: 'win' | 'loss' | 'void' = hasLoss ? 'loss' : allVoid ? 'void' : 'win'
    const wins = picks.filter(p => p.result === 'win').length

    const message = formatResultAnnouncement({
      tier: session.tier,
      date: session.date,
      comboResult,
      wins,
      total: picks.length,
      combinedOdds: session.combined_odds,
    })

    await sendMessage(message, session.telegram_msg_id ?? undefined)

    await adminSupabase
      .from('pronostic_sessions')
      .update({ combo_result: comboResult, result_posted_at: new Date().toISOString() })
      .eq('id', session.id)

    announced++
  }

  return { checked: sessions.length, announced }
}

File: backend/tools/result_checker.py
"""Port de lib/tools/result-checker.ts."""
import re
from datetime import datetime, timedelta, timezone

from supabase_client import admin_supabase
from tools.football_api import get_match_result
from tools.memory import update_performance
from tools.telegram import format_result_announcement, send_message


def _evaluate_result(bet_type: str, home_team: str, away_team: str, home_goals: int, away_goals: int) -> str:
    total = home_goals + away_goals
    bt = bet_type.lower()

    if "moins de 1.5" in bt or "under 1.5" in bt:
        return "win" if total < 1.5 else "loss"
    if "plus de 1.5" in bt or "over 1.5" in bt:
        return "win" if total > 1.5 else "loss"
    if "moins de 2.5" in bt or "under 2.5" in bt:
        return "win" if total < 2.5 else "loss"
    if "plus de 2.5" in bt or "over 2.5" in bt:
        return "win" if total > 2.5 else "loss"
    if "moins de 3.5" in bt or "under 3.5" in bt:
        return "win" if total < 3.5 else "loss"
    if "plus de 3.5" in bt or "over 3.5" in bt:
        return "win" if total > 3.5 else "loss"

    if "btts oui" in bt or "les deux équipes marquent" in bt:
        return "win" if home_goals > 0 and away_goals > 0 else "loss"
    if "btts non" in bt:
        return "win" if home_goals == 0 or away_goals == 0 else "loss"

    if "handicap" in bt:
        point_match = re.search(r"[+-]?\d+(?:\.\d+)?", bet_type)
        if not point_match:
            return "void"
        point = float(point_match.group(0))
        is_home_team = home_team.lower() in bt
        is_away_team = away_team.lower() in bt
        if not is_home_team and not is_away_team:
            return "void"
        adjusted = (home_goals + point - away_goals) if is_home_team else (away_goals + point - home_goals)
        if adjusted > 0:
            return "win"
        if adjusted < 0:
            return "loss"
        return "void"  # push — ligne entière (ex: -1, -2) tombant pile sur le score ajusté

    if "victoire" in bt and "domicile" in bt:
        return "win" if home_goals > away_goals else "loss"
    if "victoire" in bt and "extérieur" in bt:
        return "win" if away_goals > home_goals else "loss"
    if "nul" in bt or "draw" in bt:
        return "win" if home_goals == away_goals else "loss"

    # Victoire [nom équipe] — domicile si nommée, sinon void
    if bt.startswith("victoire "):
        team_name = bt.replace("victoire ", "", 1).strip()
        if team_name and home_goals > away_goals:
            return "win"
        if team_name and away_goals > home_goals:
            return "loss"

    return "void"


async def check_pending_results() -> dict:
    cutoff = (datetime.now(timezone.utc) - timedelta(hours=2)).isoformat()

    res = (
        admin_supabase.table("picks")
        .select("id, session_id, home_team, away_team, competition, match_datetime, bet_type, odds, result")
        .is_("result", "null")
        .eq("was_rejected", False)
        .lt("match_datetime", cutoff)
        .limit(50)
        .execute()
    )
    pending = res.data or []
    if not pending:
        return {"checked": 0, "updated": 0}

    updated = 0

    for pick in pending:
        match_date = pick["match_datetime"].split("T")[0]
        score = await get_match_result(pick["home_team"], pick["away_team"], match_date)
        if not score:
            continue

        result = _evaluate_result(pick["bet_type"], pick["home_team"], pick["away_team"], score["home"], score["away"])

        admin_supabase.table("picks").update(
            {"result": result, "result_checked_at": datetime.now(timezone.utc).isoformat()}
        ).eq("id", pick["id"]).execute()

        # Utilise la compétition directement depuis le pick (plus besoin de requête supplémentaire)
        competition = pick.get("competition") or "Unknown"
        update_performance(pick["bet_type"], competition, result, pick["odds"])
        updated += 1

    return {"checked": len(pending), "updated": updated}


async def announce_session_results() -> dict:
    """Poste le résultat d'un combiné sur Telegram une fois que TOUS ses picks
    sont résolus (win/loss/void) — un seul pick perdant fait perdre tout le
    combiné, comme un vrai pari combiné. Chaque session n'est annoncée
    qu'une fois (result_posted_at). Appelé après check_pending_results()
    dans le même run planifié, pour que les picks soient déjà à jour."""
    res = (
        admin_supabase.table("pronostic_sessions")
        .select("id, tier, date, combined_odds, telegram_msg_id, picks(result, was_rejected)")
        .eq("status", "published")
        .is_("result_posted_at", "null")
        .limit(20)
        .execute()
    )
    sessions = res.data or []
    if not sessions:
        return {"checked": 0, "announced": 0}

    announced = 0

    for session in sessions:
        picks = [p for p in (session.get("picks") or []) if not p.get("was_rejected")]
        if not picks:
            continue

        all_resolved = all(p.get("result") is not None for p in picks)
        if not all_resolved:
            continue

        has_loss = any(p["result"] == "loss" for p in picks)
        all_void = all(p["result"] == "void" for p in picks)
        combo_result = "loss" if has_loss else "void" if all_void else "win"
        wins = sum(1 for p in picks if p["result"] == "win")

        message = format_result_announcement(
            session["tier"], session["date"], combo_result, wins, len(picks), session.get("combined_odds")
        )

        await send_message(message, session.get("telegram_msg_id"))

        admin_supabase.table("pronostic_sessions").update(
            {"combo_result": combo_result, "result_posted_at": datetime.now(timezone.utc).isoformat()}
        ).eq("id", session["id"]).execute()

        announced += 1

    return {"checked": len(sessions), "announced": announced}
